### PR TITLE
Simplify execution engine steps

### PR DESCRIPTION
**Pull request by lptr**
The execution engine relies on `Context` and `Result` objects that carry expanding scopes of data, larger scopes typically delegating to smaller ones. This was previously done using interfaces that were implemented in-place by anonymous inner classes. However, most of that code dealt with the delegation, adding a lot of noise.

This is an attempt to simplify the code by introducing concrete classes to represent contexts and results. The code of the individual steps is significantly more readable after the changes.

The changes made some of the testing slightly more complicated. That's why `StepSpec` now has the relatively complex logic to create the stub for the context.

### CODE REVIEW

**Comment by lptr** [comment]
@bot-gradle test QFL PT please.

**Comment by lptr** [comment]
@bot-gradle test QFL PT please.

**Review by asodja — APPROVED** [review]
LGTM, logic seems simpler and less verbose. Hard for me to say if tests are that much more complicated.

We have good test coverage of that, right? :)

**Comment by lptr** [comment]
@bot-gradle test and merge please.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/steps/Result.java [hunk 1]
```diff
@@ -22,7 +22,27 @@
 
 import java.time.Duration;
 
-public interface Result {
+public class Result {
```
Comment (alllex) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/steps/Result.java:25]: Unrelated to the PR, but I would consider giving this class a slightly more concrete name like `StepResult`. I think it's good practice to use move descriptive names for public classes, when they are not nested.

**Review by alllex — APPROVED** [review]

File: subprojects/execution/src/main/java/org/gradle/internal/execution/ExecutionEngine.java [hunk 2]
```diff
@@ -83,6 +83,7 @@ interface Result {
         Optional<AfterExecutionState> getAfterExecutionState();
     }
 
+    // TOOD Make this a class
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/ExecutionEngine.java:86]: Ah yes, I meant to fix that. Actually, I wanted to remove it.

**Pull request merged**